server/src/main/java/org/elasticsearch/script/field/ValidatingMap.java

```diff
@@ -0,0 +1,259 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.script.field;
+
+import java.util.AbstractCollection;
+import java.util.AbstractMap;
+import java.util.AbstractSet;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+/**
+ * ValidatingMap is an implementation of {@code Map<String, Object>} that has a validating functions for a subset of keys.
+ *
+ * The validators are BiFunctions that should either return the identity or throw an {@link IllegalArgumentException}.  The
+ * arguments to the validator are the key and the proposed value or {@code null} for key removal.
+ *
+ * Validation occurs everywhere updates can happen, either directly via the {@link #put(String, Object)}, {@link #replace(Object, Object)},
+ * {@link #merge(Object, Object, BiFunction)} family of methods, via {@link Map.Entry#setValue(Object)},
+ * via the linked Set from {@link #entrySet()} or {@link Collection#remove(Object)} from the linked collection via {@link #values()}
+ */
+public class ValidatingMap extends AbstractMap<String, Object> {
```

Comment:
This can be package private? Or perhaps this doesn't need to be a separate class, we just need it for ingest right now, let's fold this into the real impl class and split if there is a need in the future?

Reply:
Removed as a separate class.